### ISSUE TITLE
Using sparse matrices

### ISSUE THREAD

**Issue by b-dawes** [KlugerLab/t-SNE-Heatmaps#1]
I was messing around with your tSNE heatmap code and noticed that it uses dense matrices throughout. As single-cell data is generally pretty sparse, you can get some performance increases by using sparse matrices. I was able to run the binning and summing step of the script with objects of type dgCMatrix from the Matrix library using the aggregate.Matrix function in the Matrix.utils library like so:

```R
bin_counts <- aggregate.Matrix(t(expression_matrix), cut(tsne_embedding, breaks = breaks))
bin_counts_s <- t(t(bin_counts)/rowSums(t(bin_counts)))
```

**Comment by linqiaozhi** [comment]
This is a great improvement--thanks @b-dawes!

It would be great to incorporate this into the software. Did you have a working version to pull request so that you show up as a contributor? Otherwise, I can modify it when I next update the package.

**Comment by b-dawes** [comment]
I just submitted a pull request #2 

I made it so that expression_matrix is automatically converted to a dgCMatrix. This should work for data.frames as well as regular dense matrices. I ran into some issues because aggregate.Matrix removes any empty bins so I had to add back in the empty bins. There's probably a more elegant way to do it than what I did but it seems to work.

I profiled genes_tsne_vecs with an 18,317 gene by 2,277 cell data set that was 97% sparse which I had lying around. The original code took about 22 seconds to run on my machine. Inputting a data frame into the updated version took about 0.9 seconds. Most of that time was spent converting the data frame into a dgCMatrix, when I tried inputting a dgCMatrix directly it only took 0.16 seconds so over a 100x times speed up. Hopefully you'll see similar results on your machine as well.

Best,
Brian
